controllers/topology/internal/scope/upgradetracker.go

```diff
@@ -22,31 +22,85 @@ const maxMachineDeploymentUpgradeConcurrency = 1
 
 // UpgradeTracker is a helper to capture the upgrade status and make upgrade decisions.
 type UpgradeTracker struct {
+	ControlPlane       ControlPlaneUpgradeTracker
 	MachineDeployments MachineDeploymentUpgradeTracker
 }
 
+// ControlPlaneUpgradeTracker holds the current upgrade status of the Control Plane.
+type ControlPlaneUpgradeTracker struct {
+	// PendingUpgrade is true if the control plane version needs to be updated. False otherwise.
+	PendingUpgrade bool
+
+	// IsProvisioning is true if the control plane is being provisioned for the first time. False otherwise.
+	IsProvisioning bool
+
+	// IsUpgrading is true if the control plane is in the middle of an upgrade.
+	// Note: Refer to control plane contract for definition of upgrading.
+	IsUpgrading bool
+
+	// IsScaling is true if the control plane is in the middle of a scale operation.
+	// Note: Refer to control plane contract for definition of scaling.
+	IsScaling bool
+}
+
 // MachineDeploymentUpgradeTracker holds the current upgrade status and makes upgrade
 // decisions for MachineDeployments.
 type MachineDeploymentUpgradeTracker struct {
-	names sets.String
+	deferredNames   sets.String
```

Comment:
Pending in both cases (control plane and machine deployments) is used to answer a yes or no question about "are there any changes that are on hold". 

`PendingUpgrade()` function on the MachineDeploymentUpgradeTracker and `PendingUpgrade` on the ControlPlaneUpgradeTracker have the same purpose.

Reply:
I was wondering if we should rename:
* field deferredNames => pendingNames
* func MarkDeferredUpgrade => MarkPendingUpgrade
* func DeferredUpgradeNames => PendingUpgradeNames

I think this would make it clearer that deferred and pending is the same. Currently it's e.g. a bit confusing that we add a MD to the list with `MarkDeferredUpgrade` and then we check if there is one MD in `deferredNames` with `PendingUpgrade`.

Reply:
SGTM. Will make the changes.